# 02. Предобработка данных

## Цель

Подготовить исходные таблицы Olist к дальнейшему исследовательскому анализу.

In [42]:
import pandas as pd
import numpy as np

In [98]:
orders = pd.read_csv("../data/raw/olist_orders_dataset.csv")
order_items = pd.read_csv("../data/raw/olist_order_items_dataset.csv")
order_reviews = pd.read_csv("../data/raw/olist_order_reviews_dataset.csv")
order_payments = pd.read_csv("../data/raw/olist_order_payments_dataset.csv")
products = pd.read_csv("../data/raw/olist_products_dataset.csv")
product_category_name_translation = pd.read_csv("../data/raw/product_category_name_translation.csv")
customers = pd.read_csv("../data/raw/olist_customers_dataset.csv")
sellers = pd.read_csv("../data/raw/olist_sellers_dataset.csv")
geolocation = pd.read_csv("../data/raw/olist_geolocation_dataset.csv")

### Преобразование дат

In [44]:
orders["order_purchase_timestamp"] = pd.to_datetime(orders["order_purchase_timestamp"])
orders["order_approved_at"] = pd.to_datetime(orders["order_approved_at"])
orders["order_delivered_carrier_date"] = pd.to_datetime(orders["order_delivered_carrier_date"])
orders["order_delivered_customer_date"] = pd.to_datetime(orders["order_delivered_customer_date"])
orders["order_estimated_delivery_date"] = pd.to_datetime(orders["order_estimated_delivery_date"])

order_items["shipping_limit_date"] = pd.to_datetime(order_items["shipping_limit_date"])
order_reviews["review_creation_date"] = pd.to_datetime(order_reviews["review_creation_date"])
order_reviews["review_answer_timestamp"] = pd.to_datetime(order_reviews["review_answer_timestamp"])

In [45]:
orders.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  str           
 1   customer_id                    99441 non-null  str           
 2   order_status                   99441 non-null  str           
 3   order_purchase_timestamp       99441 non-null  datetime64[us]
 4   order_approved_at              99281 non-null  datetime64[us]
 5   order_delivered_carrier_date   97658 non-null  datetime64[us]
 6   order_delivered_customer_date  96476 non-null  datetime64[us]
 7   order_estimated_delivery_date  99441 non-null  datetime64[us]
dtypes: datetime64[us](5), str(3)
memory usage: 6.1 MB


In [46]:
order_items.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  str           
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  str           
 3   seller_id            112650 non-null  str           
 4   shipping_limit_date  112650 non-null  datetime64[us]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[us](1), float64(2), int64(1), str(3)
memory usage: 6.0 MB


In [47]:
order_reviews.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  str           
 1   order_id                 99224 non-null  str           
 2   review_score             99224 non-null  int64         
 3   review_comment_title     11568 non-null  str           
 4   review_comment_message   40977 non-null  str           
 5   review_creation_date     99224 non-null  datetime64[us]
 6   review_answer_timestamp  99224 non-null  datetime64[us]
dtypes: datetime64[us](2), int64(1), str(4)
memory usage: 5.3 MB


In [48]:
orders[orders["order_approved_at"].isna()]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
1130,00b1cb0320190ca0daa2c88b35206009,3532ba38a3fd242259a514ac2b6ae6b6,canceled,2018-08-28 15:26:39,NaT,NaT,NaT,2018-09-12
1801,ed3efbd3a87bea76c2812c66a0b32219,191984a8ba4cbb2145acb4fe35b69664,canceled,2018-09-20 13:54:16,NaT,NaT,NaT,2018-10-17
1868,df8282afe61008dc26c6c31011474d02,aa797b187b5466bc6925aaaa4bb3bed1,canceled,2017-03-04 12:14:30,NaT,NaT,NaT,2017-04-10
2029,8d4c637f1accf7a88a4555f02741e606,b1dd715db389a2077f43174e7a675d07,canceled,2018-08-29 16:27:49,NaT,NaT,NaT,2018-09-13
2161,7a9d4c7f9b068337875b95465330f2fc,7f71ae48074c0cfec9195f88fcbfac55,canceled,2017-05-01 16:12:39,NaT,NaT,NaT,2017-05-30
...,...,...,...,...,...,...,...,...
97696,5a00b4d35edffc56b825c3646a99ba9d,6a3bdf004ca96338fb5fad1b8d93c2e6,canceled,2017-07-02 15:38:46,NaT,NaT,NaT,2017-07-25
98415,227c804e2a44760671a6a5697ea549e4,62e7477e75e542243ee62a0ba73f410f,canceled,2017-09-28 15:02:56,NaT,NaT,NaT,2017-10-16
98909,e49e7ce1471b4693482d40c2bd3ad196,e4e7ab3f449aeb401f0216f86c2104db,canceled,2018-08-07 11:16:28,NaT,NaT,NaT,2018-08-10
99283,3a3cddda5a7c27851bd96c3313412840,0b0d6095c5555fe083844281f6b093bb,canceled,2018-08-31 16:13:44,NaT,NaT,NaT,2018-10-01


### Результат преобразования дат

- Все временные столбцы в таблицах `orders`, `order_items` и `order_reviews` успешно преобразованы из строкового типа в тип `datetime`.
- Пропущенные значения в столбцах дат таблицы `orders` сохранились как `NaT`, что является корректным представлением отсутствующей даты в pandas.
- Теперь столбцы можно использовать для вычисления длительности доставки, задержек и других временных признаков.

### Проверка невозможных значений

На данном этапе проверяются числовые признаки на наличие физически невозможных или некорректных значений (отрицательных цен, отрицательной стоимости доставки, нулевого веса и других потенциальных аномалий).

In [49]:
order_items[order_items["price"] <= 0]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value


In [50]:
order_items[order_items["freight_value"] < 0]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value


In [51]:
order_payments[order_payments["payment_value"] <= 0]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
19922,8bcbe01d44d147f901cd3192671144db,4,voucher,1,0.0
36822,fa65dad1b0e818e3ccc5cb0e39231352,14,voucher,1,0.0
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.0
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.0
77885,fa65dad1b0e818e3ccc5cb0e39231352,13,voucher,1,0.0
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0
100766,b23878b3e8eb4d25a158f57d96331b18,4,voucher,1,0.0


In [52]:
order_payments["payment_type"].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

In [53]:
order_payments[
    order_payments["payment_value"] == 0
]["payment_type"].value_counts()

payment_type
voucher        6
not_defined    3
Name: count, dtype: int64

In [54]:
products[products["product_height_cm"] <= 0]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm


In [55]:
products[products["product_length_cm"] <= 0]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm


In [56]:
products[products["product_width_cm"] <= 0]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm


In [57]:
products[products["product_weight_g"] <= 0]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
9769,81781c0fed9fe1ad6e8c81fca1e1cb08,cama_mesa_banho,51.0,529.0,1.0,0.0,30.0,25.0,30.0
13683,8038040ee2a71048d4bdbbdc985b69ab,cama_mesa_banho,48.0,528.0,1.0,0.0,30.0,25.0,30.0
14997,36ba42dd187055e1fbe943b2d11430ca,cama_mesa_banho,53.0,528.0,1.0,0.0,30.0,25.0,30.0
32079,e673e90efa65a5409ff4196c038bb5af,cama_mesa_banho,53.0,528.0,1.0,0.0,30.0,25.0,30.0


In [58]:
product_category_name_translation[product_category_name_translation["product_category_name"] == "cama_mesa_banho"]

,product_category_name,product_category_name_english
3,cama_mesa_banho,bed_bath_table


#### Проверка корректности сумм платежей

Проверка показала отсутствие отрицательных значений в столбце `payment_value`.

Обнаружено 9 записей с нулевой суммой платежа (`payment_value = 0`). Большинство из них относятся к типу оплаты `voucher`, а часть — к типу `not_defined`. Такие записи встречаются крайне редко и, вероятно, отражают особенности формирования данных, а не ошибки. На данном этапе они сохраняются без изменений.

#### Проверка физических характеристик товаров

При проверке столбца `product_weight_g` обнаружено 4 товара с весом 0 г. Все они относятся к категории `cama_mesa_banho` (`bed_bath_table`) и имеют заполненные остальные характеристики (размеры, описание, количество фотографий).

Поскольку физический товар не может иметь нулевой вес, данные записи, вероятно, являются ошибками или пропущенными значениями, записанными как `0`. На данном этапе они будут сохранены и рассмотрены при дальнейшей обработке данных.

### Проверка дат на логичность

На данном этапе проверяются даты из таблицы `orders` на логичную последовательность: покупка, подтверждение, передача перевозчику, доставка клиенту.

In [59]:
orders[orders["order_purchase_timestamp"] > orders["order_approved_at"]]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date


In [60]:
orders[orders["order_approved_at"] > orders["order_delivered_carrier_date"]]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
15,dcb36b511fcac050b97cd5c05de84dc3,3b6828a50ffe546942b7a473d70ac0fc,delivered,2018-06-07 19:03:12,2018-06-12 23:31:02,2018-06-11 14:54:00,2018-06-21 15:34:32,2018-07-04
64,688052146432ef8253587b930b01a06d,81e08b08e5ed4472008030d70327c71f,delivered,2018-04-22 08:48:13,2018-04-24 18:25:22,2018-04-23 19:19:14,2018-04-24 19:31:58,2018-05-15
199,58d4c4747ee059eeeb865b349b41f53a,1755fad7863475346bc6c3773fe055d3,delivered,2018-07-21 12:49:32,2018-07-26 23:31:53,2018-07-24 12:57:00,2018-07-25 23:58:19,2018-07-31
210,412fccb2b44a99b36714bca3fef8ad7b,c6865c523687cb3f235aa599afef1710,delivered,2018-07-22 22:30:05,2018-07-23 12:31:53,2018-07-23 12:24:00,2018-07-24 19:26:42,2018-07-31
415,56a4ac10a4a8f2ba7693523bb439eede,78438ba6ace7d2cb023dbbc81b083562,delivered,2018-07-22 13:04:47,2018-07-27 23:31:09,2018-07-24 14:03:00,2018-07-28 00:05:39,2018-08-06
...,...,...,...,...,...,...,...,...
99091,240ead1a7284667e0ec71d01f80e4d5e,fcdd7556401aaa1c980f8b67a69f95dc,delivered,2018-07-02 16:30:02,2018-07-05 16:17:59,2018-07-05 14:11:00,2018-07-10 23:21:47,2018-07-24
99230,78008d03bd8ef7fcf1568728b316553c,043e3254e68daf7256bda1c9c03c2286,delivered,2018-07-03 13:11:13,2018-07-05 16:32:52,2018-07-03 12:57:00,2018-07-10 17:47:39,2018-07-23
99266,76a948cd55bf22799753720d4545dd2d,3f20a07b28aa252d0502fe7f7eb030a9,delivered,2018-01-30 02:41:30,2018-02-04 23:31:46,2018-01-31 18:11:58,2018-03-18 20:08:50,2018-03-02
99377,a6bd1f93b7ff72cc348ca07f38ec4bee,6d63fa86bd2f62908ad328325799152f,delivered,2018-04-20 17:28:40,2018-04-24 19:26:10,2018-04-23 17:18:40,2018-04-28 17:38:42,2018-05-15


In [61]:
delta = (
    orders["order_approved_at"] - orders["order_delivered_carrier_date"]
)
delta[delta > pd.Timedelta(0)].describe()

count                      1359
mean     1 days 00:45:07.153053
std      4 days 19:18:28.055754
min             0 days 00:00:21
25%      0 days 01:24:55.500000
50%             0 days 17:10:04
75%             1 days 01:57:24
max           171 days 05:15:22
dtype: object

In [67]:
delta.sort_values(ascending=False).head(10)

25883   171 days 05:15:22
14562     9 days 08:55:48
46163     9 days 08:11:25
98710     9 days 03:55:56
41592     9 days 02:45:45
11738     9 days 01:51:58
85393     9 days 01:21:40
31211     9 days 01:10:54
55302     9 days 01:06:08
68315     8 days 10:05:24
dtype: timedelta64[us]

In [64]:
orders.loc[
    delta.sort_values(ascending=False).head(10).index,
    [
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date"
    ]
]

,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date
25883,2018-07-16 18:40:53,2018-07-16 18:50:22,2018-01-26 13:35:00,2018-07-23 20:04:45
14562,2017-09-01 19:04:22,2017-09-13 22:06:11,2017-09-04 13:10:23,2017-09-08 20:13:03
46163,2017-09-01 20:04:28,2017-09-13 22:17:15,2017-09-04 14:05:50,2017-09-09 15:12:44
98710,2017-09-01 20:28:02,2017-09-13 22:03:51,2017-09-04 18:07:55,2017-09-13 22:24:46
41592,2017-09-01 18:40:44,2017-09-13 21:58:04,2017-09-04 19:12:19,2017-09-08 20:07:45
11738,2017-09-01 18:45:33,2017-09-13 22:04:39,2017-09-04 20:12:41,2017-09-11 14:15:02
85393,2017-09-01 20:05:55,2017-09-13 21:58:38,2017-09-04 20:36:58,2017-09-08 20:41:16
31211,2017-09-01 20:05:42,2017-09-13 22:00:51,2017-09-04 20:49:57,2017-09-06 22:15:45
55302,2017-09-01 18:49:54,2017-09-13 21:56:08,2017-09-04 20:50:00,2017-09-30 17:33:43
68315,2017-09-01 20:17:09,2017-09-13 22:03:42,2017-09-05 11:58:18,2017-09-18 17:52:20


Проверка логической последовательности дат выявила 1359 заказов, у которых дата подтверждения заказа `order_approved_at` позже даты передачи перевозчику    order_delivered_carrier_date    . Для большинства записей нарушение составляет менее суток, однако встречаются отдельные экстремальные случаи `до 171 дня`, которые с высокой вероятностью являются ошибками исходных данных. На данном этапе изменения в таблицу не вносятся. Решение об обработке таких записей будет принято после завершения исследовательского анализа и оценки их влияния на результаты.

In [73]:
orders[orders["order_delivered_carrier_date"] > orders["order_delivered_customer_date"]]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
6437,a1abeb653a4d4cd1e142ccb8c82cd069,5f50465da00b7fed5dd1239f4ecf6e2c,delivered,2017-07-20 11:20:52,2017-07-21 06:43:14,2017-07-28 16:57:58,2017-07-25 19:32:56,2017-08-14
9553,383aa8b2724fe452d9ccd9934a8c628b,b1cb2f9d7a19480f3749e248db14d58f,delivered,2017-07-02 20:58:43,2017-07-02 21:10:20,2017-07-07 17:22:41,2017-07-06 14:27:51,2017-07-21
13487,cb1134f9010d242e9515ad1c78ec0c39,2fd33ac77677bd214b1882868317eeed,delivered,2017-07-16 12:35:34,2017-07-18 06:03:50,2017-07-20 19:22:02,2017-07-19 14:13:28,2017-08-08
14474,dceb62e8fa94b46006c9554fed743df0,2721900eb4e0f1cc2c836dd7bc1b1e11,delivered,2017-07-20 20:58:05,2017-07-22 11:45:11,2017-08-01 18:23:30,2017-07-26 18:09:10,2017-08-11
19268,5f9d46795c3126674e52becb3a1a517f,79287bcaafdde5c793b996fc40bb7d9f,delivered,2017-07-18 11:48:20,2017-07-18 12:03:29,2017-07-20 23:03:42,2017-07-20 18:52:41,2017-07-31
21338,8c78d01de3a9009e23d6877a7cc9be20,6cd7106899e59a1fbd0622d5f1efedf4,delivered,2016-10-08 15:36:50,2016-10-08 18:13:44,2016-10-26 11:41:53,2016-10-25 17:51:46,2016-11-30
22520,b27af682321527a6349f1761eb3f360c,9859dd92e872dbaa60ca3cd5f0d7ad07,delivered,2017-06-14 20:17:04,2017-06-14 20:30:08,2017-06-27 14:51:54,2017-06-26 15:45:35,2017-07-14
25393,1cc3ae63caffff2d6c3ee3e78e074acf,01c843a2c0600def0b7693dba47af460,delivered,2017-08-07 21:35:22,2017-08-08 21:45:15,2017-08-10 18:28:56,2017-08-10 18:05:38,2017-08-25
25646,e37f11cae9985ca58f0b56f268720537,3947a361301f2ff0f3223159a0f2701c,delivered,2017-07-26 11:46:34,2017-07-27 10:10:16,2017-08-01 18:17:47,2017-07-31 17:49:56,2017-08-24
27470,fa3e37584f4fdb1ded0e0de700dfcb4e,63be4feff10a0b1d85f2cfbf10df9754,delivered,2017-07-30 19:32:23,2017-07-30 19:45:09,2017-08-09 18:18:43,2017-08-01 21:13:01,2017-08-18


In [74]:
orders[orders["order_delivered_carrier_date"] > orders["order_delivered_customer_date"]].shape

(23, 8)

In [71]:
delta = (
    orders["order_delivered_carrier_date"] - orders["order_delivered_customer_date"]
)

delta.sort_values(ascending=False).head(10)

34939   16 days 02:18:29
27470    7 days 21:05:42
45302    7 days 03:58:13
14474    6 days 00:14:20
49933    5 days 17:39:49
71227    5 days 16:34:11
74967    5 days 00:56:54
41636    4 days 22:38:22
6437     2 days 21:25:02
78556    1 days 23:38:21
dtype: timedelta64[us]

При проверке последовательности этапов доставки обнаружено 23 заказа `≈0,02% от общего числа`, у которых дата передачи заказа перевозчику `order_delivered_carrier_date` позже даты доставки клиенту `order_delivered_customer_date`. Такие записи противоречат ожидаемой логике процесса доставки и, вероятнее всего, являются ошибками исходных данных. Поскольку доля подобных случаев крайне мала и невозможно однозначно определить некорректное значение, на данном этапе записи сохраняются без изменений.

Проверка отрицательной длительности доставки

In [75]:
delivery_time = (
    orders["order_delivered_customer_date"] -
    orders["order_purchase_timestamp"]
)

delivery_time[delivery_time < pd.Timedelta(0)]

Series([], dtype: timedelta64[us])

In [76]:
orders[
    (orders["order_status"] == "delivered") &
    (orders["order_delivered_customer_date"].isna())
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaT,2017-12-18
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaT,2018-07-16
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaT,2018-07-30
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaT,2018-07-30
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaT,2018-07-24
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaT,2018-06-26
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaT,2018-07-19


Обнаружены заказы со статусом delivered, у которых отсутствует дата доставки клиенту (order_delivered_customer_date). Такие записи противоречат ожидаемой логике данных, однако их количество крайне мало по сравнению с общим объёмом датасета. На данном этапе они сохраняются без изменений и будут учитываться при последующем анализе.

In [77]:
orders[
    (orders["order_status"] != "delivered") &
    (~orders["order_delivered_customer_date"].isna())
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
2921,1950d777989f6a877539f53795b4c3c3,1bccb206de9f0f25adc6871a1bcf77b2,canceled,2018-02-19 19:48:52,2018-02-19 20:56:05,2018-02-20 19:57:13,2018-03-21 22:03:51,2018-03-09
8791,dabf2b0e35b423f94618bf965fcb7514,5cdec0bb8cbdf53ffc8fdc212cd247c6,canceled,2016-10-09 00:56:52,2016-10-09 13:36:58,2016-10-13 13:36:59,2016-10-16 14:36:59,2016-11-30
58266,770d331c84e5b214bd9dc70a10b829d0,6c57e6119369185e575b36712766b0ef,canceled,2016-10-07 14:52:30,2016-10-07 15:07:10,2016-10-11 15:07:11,2016-10-14 15:07:11,2016-11-29
59332,8beb59392e21af5eb9547ae1a9938d06,bf609b5741f71697f65ce3852c5d2623,canceled,2016-10-08 20:17:50,2016-10-09 14:34:30,2016-10-14 22:45:26,2016-10-19 18:47:43,2016-11-30
92636,65d1e226dfaeb8cdc42f665422522d14,70fc57eeae292675927697fe03ad3ff5,canceled,2016-10-03 21:01:41,2016-10-04 10:18:57,2016-10-25 12:14:28,2016-11-08 10:58:34,2016-11-25
94399,2c45c33d2f9cb8ff8b1c86cc28c11c30,de4caa97afa80c8eeac2ff4c8da5b72e,canceled,2016-10-09 15:39:56,2016-10-10 10:40:49,2016-10-14 10:40:50,2016-11-09 14:53:50,2016-12-08


При проверке согласованности полей `order_status` и `order_delivered_customer_date` обнаружено 6 заказов со статусом `canceled`, для которых при этом указана дата доставки клиенту. Такие записи противоречат ожидаемой логике данных и, вероятно, связаны с особенностями формирования исходной базы данных или редкими ошибками. Поскольку их количество составляет менее 0.01% от общего объёма данных, на данном этапе они сохраняются без изменений.

### Анализ пропусков

#### Пропуски в orders

In [79]:
orders[orders["order_approved_at"].isna()]["order_status"].value_counts()

order_status
canceled     141
delivered     14
created        5
Name: count, dtype: int64

Большинство пропусков приходится на отменённые (canceled) и ещё не обработанные (created) заказы.

Также обнаружено 14 заказов со статусом delivered без даты подтверждения. Это небольшое количество записей, которые, вероятно, являются особенностями данных или ошибками заполнения. На данном этапе они сохраняются.

In [80]:
orders[orders["order_delivered_carrier_date"].isna()]["order_status"].value_counts()

order_status
unavailable    609
canceled       550
invoiced       314
processing     301
created          5
approved         2
delivered        2
Name: count, dtype: int64

Пропуски в основном соответствуют заказам, которые не были переданы перевозчику (processing, invoiced, unavailable, canceled).

Обнаружены лишь две записи со статусом delivered без даты передачи перевозчику. Из-за крайне малого количества они остаются без изменений.

In [81]:
orders[orders["order_delivered_customer_date"].isna()]["order_status"].value_counts()

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64

Большинство пропусков наблюдается у заказов, которые ещё не были доставлены клиенту либо были отменены.

Лишь восемь заказов имеют статус delivered при отсутствии даты доставки. Эти записи сохраняются для дальнейшего анализа.

#### Пропуски в order_reviews

Пропуски в review_comment_title и
review_comment_message означают,
что пользователь оставил только числовую оценку.

Это не ошибка данных.

Записи сохраняются.

#### Пропуски в products

Дополнительный анализ показал, что пропуски в таблице `products` имеют систематический характер.

Все 610 записей без значения `product_category_name` одновременно содержат пропуски в столбцах `product_name_length`, `product_description_length` и `product_photos_qty`, что указывает на отсутствие части информации о товаре, а не на случайную потерю отдельных значений. При этом физические характеристики товаров (вес и размеры) в большинстве таких записей заполнены.

Пропуски в столбцах физических характеристик (`product_weight_g`, `product_length_cm`, `product_height_cm`, `product_width_cm`) встречаются лишь в двух записях для каждого признака, поэтому не оказывают существенного влияния на дальнейший анализ.

На данном этапе пропущенные значения не заполняются и не удаляются. Они сохраняются в исходном виде, поскольку их доля невелика, а искусственное заполнение могло бы исказить исходные данные. При необходимости обработка этих записей будет выполнена на этапе анализа или построения моделей.

### Проверка связи между таблицами

In [90]:
set(order_items["product_id"]) - set(products["product_id"])

set()

In [96]:
set(orders["customer_id"]) - set(customers["customer_id"])

set()

In [91]:
set(order_items["order_id"]) - set(orders["order_id"])

set()

In [92]:
set(order_payments["order_id"]) - set(orders["order_id"])

set()

In [ ]:
set(order_reviews["order_id"]) - set(orders["order_id"])

set()

In [97]:
set(order_items["seller_id"]) - set(sellers["seller_id"])

set()

### Вывод

Проверка связей между таблицами показала отсутствие нарушений ссылочной целостности. Все идентификаторы товаров, заказов, покупателей, продавцов, платежей и отзывов корректно связаны между собой. Записей, ссылающихся на несуществующие объекты в связанных таблицах, не обнаружено.

### Сохранение обработанных таблиц

In [99]:
orders.to_csv("../data/processed/orders.csv", index=False)
customers.to_csv("../data/processed/customers.csv", index=False)
order_items.to_csv("../data/processed/order_items.csv", index=False)
order_payments.to_csv("../data/processed/order_payments.csv", index=False)
order_reviews.to_csv("../data/processed/order_reviews.csv", index=False)
products.to_csv("../data/processed/products.csv", index=False)
sellers.to_csv("../data/processed/sellers.csv", index=False)
geolocation.to_csv("../data/processed/geolocation.csv", index=False)
product_category_name_translation.to_csv(
    "../data/processed/product_category_name_translation.csv",
    index=False
)

## Итоги этапа предобработки

В ходе предобработки данных были выполнены следующие действия:

- даты приведены к типу `datetime`;
- проверены и проанализированы невозможные значения;
- выполнена проверка логической последовательности дат;
- исследованы пропущенные значения и принято решение сохранить их без изменений, поскольку большинство пропусков объясняется бизнес-логикой;
- проверена ссылочная целостность между таблицами — нарушений не обнаружено.

По итогам проверки существенных проблем, требующих удаления или корректировки большого количества записей, выявлено не было. Все изменения на данном этапе ограничились приведением типов данных, а также анализом качества данных без искусственного изменения исходной информации.

Подготовленные таблицы сохранены в папке `data/processed` и готовы к дальнейшему исследовательскому анализу.